r/src/arrow_cpp11.h

```diff
@@ -157,8 +157,15 @@ struct ns {
 
 template <typename Pointer>
 Pointer r6_to_pointer(SEXP self) {
-  return reinterpret_cast<Pointer>(
-      R_ExternalPtrAddr(Rf_findVarInFrame(self, arrow::r::symbols::xp)));
+  void* p = R_ExternalPtrAddr(Rf_findVarInFrame(self, arrow::r::symbols::xp));
+  if (p == nullptr) {
+    SEXP klass = Rf_getAttrib(self, R_ClassSymbol);
+    std::string first_class(Rf_isNull(klass) ? "ArrowObject"
+                                             : CHAR(STRING_ELT(klass, 0)));
```

Comment:
1) if this is an R6 object, how can class ever be NULL?

2) Why not take the last element in `klass`, i.e. the actual subclass?

Reply:
I'll refine about the NULL case, but the class we do want is the first one: 

``` r
library(arrow, warn.conflicts = FALSE)

a <- Array$create(1:10)
class(a)
#> [1] "Array"       "ArrowObject" "R6"
```

<sup>Created on 2020-09-24 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>